File: exercises/RB101_small_problems/easy_8/all_substrings.rb
=begin

P: Write a method that takes a string and returns all
   substrings in array format. Order of the substrings
   should be with respective to their index, and shortest -> longest.

E: - find all substrings
   - order of the substrings must shortest -> longest

I: - none identified

T: ('abcde') == [
  'a', 'ab', 'abc', 'abcd', 'abcde',
  'b', 'bc', 'bcd', 'bcde',
  'c', 'cd', 'cde',
  'd', 'de',
  'e'
]

A: /* given a string */
   - initialize substrings array
   - starting from index 0 find all substrings
     - once all substrings have been found from index 0
       - find substrings from index 1 and so on...
     - push substrings to substrings array
   - return array

=end

def substrings(str)
  substrings = []
  (0...str.size).each do |i|
    (i...str.size).each do |j|
      substrings << str[i..j]
    end
  end
  substrings
end

p substrings('abcde') == [
  'a', 'ab', 'abc', 'abcd', 'abcde',
  'b', 'bc', 'bcd', 'bcde',
  'c', 'cd', 'cde',
  'd', 'de',
  'e'
]

File: exercises/JS210_Small_Problems/List_Processing/allSubstrings.js
// Solution 1
function substrings(string) {
  let result = [];

  for (let i = 0; i < string.length; i += 1) {
    for (let j = 1; j <= string.length; j += 1) {
      if (string.slice(i, j) !== '') {
        result.push(string.slice(i, j));
      }
    }
  }
  return result;
}

// Solution 2
function leadingSubstrings(str) {
  return [...str].map((_, idx) => str.substring(0, idx + 1))
}

function substrings(str) {
  return [...str].flatMap((_, idx) => leadingSubstrings(str.slice(idx)));
}

console.log(substrings('abcde'));

// returns
[ "a", "ab", "abc", "abcd", "abcde",
  "b", "bc", "bcd", "bcde",
  "c", "cd", "cde",
  "d", "de",
  "e" ]